server/extension/extension.go

```diff
@@ -277,81 +402,106 @@ func (m *Manager) registerExtensions(r *mux.Router, extConfigs *ExtensionConfigs
+	}
+
+	// just retrieve the app after checking if subject has access to it
+	app, err := m.application.Get(rr.ApplicationNamespace, rr.ApplicationName)
+	if err != nil {
+		return nil, fmt.Errorf("error getting application: %s", err)
+	}
+	if app == nil {
+		return nil, fmt.Errorf("invalid Application provided in the %q header", HeaderArgoCDApplicationName)
+	}
+
+	// if resources are provided in the HeaderArgoCDResourceGVKName
+	// it will validate if it belongs to the application
+	for _, resource := range rr.Resources {
+		found := false
+		for _, child := range app.Status.Resources {
+			if child.GroupVersionKind() == resource.Gvk &&
+				child.Name == resource.Name {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return nil, fmt.Errorf("resource from header %q does not belong to the application", HeaderArgoCDResourceGVKName)
+		}
+	}
+
+	return app, nil
+}
+
 // CallExtension returns a handler func responsible for forwarding requests to the
 // extension service. The request will be sanitized by removing sensitive headers.
 func (m *Manager) CallExtension(extName string, proxyByCluster map[string]*httputil.ReverseProxy) func(http.ResponseWriter, *http.Request) {
 	return func(w http.ResponseWriter, r *http.Request) {
-		sanitizeRequest(r, extName)
-		if len(proxyByCluster) == 1 {
-			for _, proxy := range proxyByCluster {
-				proxy.ServeHTTP(w, r)
-				return
-			}
-		}
-		appHeader := r.Header.Get(HeaderArgoCDApplicationName)
-		if appHeader == "" {
-			msg := fmt.Sprintf("Header %q must be provided", HeaderArgoCDApplicationName)
-			m.writeErrorResponse(http.StatusBadRequest, msg, w)
-			return
-		}
-		appNamespace, appName, err := getAppName(appHeader)
+		reqResources, err := ValidateHeaders(r)
 		if err != nil {
-			msg := fmt.Sprintf("Error getting application name: %s", err)
-			m.writeErrorResponse(http.StatusBadRequest, msg, w)
+			http.Error(w, fmt.Sprintf("Invalid headers: %s", err), http.StatusBadRequest)
 			return
 		}
-		app, err := m.application.Get(appNamespace, appName)
+		app, err := m.authorize(r.Context(), reqResources, extName)
 		if err != nil {
-			msg := fmt.Sprintf("Error getting application: %s", err)
-			m.writeErrorResponse(http.StatusBadRequest, msg, w)
+			http.Error(w, fmt.Sprintf("Unauthorized extension request: %s", err), http.StatusUnauthorized)
 			return
 		}
-		if app == nil {
-			msg := fmt.Sprintf("Invalid Application: %s", appHeader)
-			m.writeErrorResponse(http.StatusBadRequest, msg, w)
-			return
+
+		sanitizeRequest(r, extName)
+
+		// This is the case where there is only one proxy configured
+		// for this extension.
+		if len(proxyByCluster) == 1 {
+			for _, proxy := range proxyByCluster {
+				// in this case we just forward the request to the single
+				// proxy and return
+				proxy.ServeHTTP(w, r)
+				return
+			}
 		}
 		clusterName := app.Spec.Destination.Name
 		if clusterName == "" {
 			clusterName = app.Spec.Destination.Server
 		}
```

Comment:
We should probably document that we don't automatically match cluster names to server URLs.

Reply:
done